src/hotspot/share/opto/mulnode.cpp

```diff
@@ -1683,3 +1713,58 @@ const Type* RotateRightNode::Value(PhaseGVN* phase) const {
     return TypeLong::LONG;
   }
 }
+
+bool MulNode::AndIL_shift_and_mask(PhaseGVN* phase, Node* mask, Node* shift, BasicType bt) const {
+  if (mask == NULL || shift == NULL) {
+    return false;
+  }
+  const TypeInteger* mask_t = phase->type(mask)->isa_integer(bt);
+  const TypeInteger* shift_t = phase->type(shift)->isa_integer(bt);
+  if (mask_t == NULL || shift_t == NULL) {
+    return false;
+  }
+  if (bt == T_LONG && shift != NULL && shift->Opcode() == Op_ConvI2L) {
+    bt = T_INT;
+    shift = shift->in(1);
+    if (shift == NULL) {
+      return false;
+    }
+  }
+  if (shift->Opcode() != Op_LShift(bt)) {
+    return false;
+  }
+  Node* shift2 = shift->in(2);
+  if (shift2 == NULL) {
+    return false;
+  }
+  const Type* shift2_t = phase->type(shift2);
+  if (!shift2_t->isa_int() || !shift2_t->is_int()->is_con()) {
+    return false;
+  }
+
+  jint shift_con = shift2_t->is_int()->get_con() & ((bt == T_INT ? BitsPerJavaInteger : BitsPerJavaLong) - 1);
+  if ((1L << shift_con) > mask_t->hi_as_long() && mask_t->lo_as_long() >= 0) {
+    return true;
+  }
+
+  return false;
+}
+
+Node* MulNode::AndIL_add_shift_and_mask(PhaseGVN* phase, BasicType bt) {
+  Node* in1 = in(1);
+  Node* in2 = in(2);
```

Comment:
Caller already determine that `in2` is `const int mask = t2->get_con()`. You can pass it here.